File: components/accordion/android/CiataAccordion.kt
package org.ciata.ds.accordion

import androidx.compose.foundation.layout.Column
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.semantics.semantics

/** Implementação experimental do CMP-0014 Accordion para Jetpack Compose. */
@Composable
fun CiataAccordion(
    labels: List<String>,
    expanded: Set<Int>,
    onToggle: (Int) -> Unit,
    content: @Composable (Int) -> Unit,
) {
    val normalizedLabels = labels.map(String::trim)

    require(normalizedLabels.isNotEmpty()) { "labels não pode ser vazio." }
    require(normalizedLabels.all { it.isNotEmpty() }) { "rótulos não podem ser vazios." }
    require(expanded.all { it in normalizedLabels.indices }) { "expanded contém índice inválido." }

    Column {
        normalizedLabels.forEachIndexed { index, label ->
            val isExpanded = index in expanded
            TextButton(
                onClick = { onToggle(index) },
                modifier = Modifier.semantics {
                    stateDescription = if (isExpanded) "Expandido" else "Recolhido"
                },
            ) {
                Text(label)
            }
            if (isExpanded) {
                content(index)
            }
        }
    }
}

File: components/accordion/ios/CiataAccordion.swift
import SwiftUI

/// Implementação experimental do CMP-0014 Accordion para SwiftUI.
public struct CiataAccordion<Content: View>: View {
    private let labels: [String]
    @Binding private var expanded: Set<Int>
    private let content: (Int) -> Content

    public init(
        labels: [String],
        expanded: Binding<Set<Int>>,
        @ViewBuilder content: @escaping (Int) -> Content
    ) {
        let normalizedLabels = labels.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }

        precondition(!normalizedLabels.isEmpty, "labels não pode ser vazio")
        precondition(normalizedLabels.allSatisfy { !$0.isEmpty }, "rótulos não podem ser vazios")
        precondition(expanded.wrappedValue.allSatisfy { normalizedLabels.indices.contains($0) }, "expanded contém índice inválido")
        self.labels = normalizedLabels
        self._expanded = expanded
        self.content = content
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            ForEach(labels.indices, id: \.self) { index in
                DisclosureGroup(
                    isExpanded: Binding(
                        get: { expanded.contains(index) },
                        set: { isOpen in
                            if isOpen { expanded.insert(index) }
                            else { expanded.remove(index) }
                        }
                    ),
                    content: { content(index) },
                    label: { Text(labels[index]) }
                )
            }
        }
    }
}
